controllers/propagator/replicatedpolicy_setup.go

```diff
@@ -1,48 +1,76 @@
 package propagator
 
 import (
+	"context"
+	"sync"
+	"time"
+
 	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/event"
 	"sigs.k8s.io/controller-runtime/pkg/handler"
 	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 	"sigs.k8s.io/controller-runtime/pkg/source"
 
 	policiesv1 "open-cluster-management.io/governance-policy-propagator/api/v1"
 	"open-cluster-management.io/governance-policy-propagator/controllers/common"
 )
 
-func (r *ReplicatedPolicyReconciler) SetupWithManager(mgr ctrl.Manager, additionalSources ...source.Source) error {
-	builder := ctrl.NewControllerManagedBy(mgr).
+func (r *ReplicatedPolicyReconciler) SetupWithManager(
+	mgr ctrl.Manager, dynWatcherSrc source.Source, updateSrc source.Source,
+) error {
+	return ctrl.NewControllerManagedBy(mgr).
 		Named("replicated-policy").
 		For(
 			&policiesv1.Policy{},
-			builder.WithPredicates(replicatedPolicyPredicates()),
-		)
-
-	for _, source := range additionalSources {
-		builder.WatchesRawSource(source, &handler.EnqueueRequestForObject{})
-	}
-
-	return builder.Complete(r)
+			builder.WithPredicates(replicatedPolicyPredicates(r.ResourceVersions))).
+		WatchesRawSource(
+			dynWatcherSrc,
+			// The dependency-watcher could create an event before the same sort of watch in the
+			// controller-runtime triggers an update in the cache. This tries to ensure the cache is
+			// updated before the reconcile is triggered.
+			&delayGeneric{
+				EventHandler: &handler.EnqueueRequestForObject{},
+				delay:        time.Second * 3,
+			}).
+		WatchesRawSource(
+			updateSrc,
+			&handler.EnqueueRequestForObject{}).
+		Complete(r)
 }
 
 // replicatedPolicyPredicates triggers reconciliation if the policy is a replicated policy, and is
-// not a pure status update.
-func replicatedPolicyPredicates() predicate.Funcs {
+// not a pure status update. It will use the ResourceVersions cache to try and skip events caused
+// by the replicated policy reconciler itself.
+func replicatedPolicyPredicates(resourceVersions *sync.Map) predicate.Funcs {
 	return predicate.Funcs{
 		CreateFunc: func(e event.CreateEvent) bool {
 			_, isReplicated := e.Object.GetLabels()[common.RootPolicyLabel]
+			if !isReplicated {
+				return false
+			}
 
-			// NOTE: can we ignore the create event from when this controller creates the resource?
-			return isReplicated
+			version := safeLoad(resourceVersions, e.Object)
+			if version == "creating" || version == e.Object.GetResourceVersion() {
```

Comment:
If you follow my previous comment of adding this ignore logic in the reconciler itself, then the value should always be an actual resource version except for the delete case, which would have an empty string.